devlib/utils/ssh.py

```diff
@@ -299,6 +312,469 @@ def __init__(self,
         self.conn = ssh_get_shell(host, username, password, None, port, timeout, True, original_prompt)
 
 
+class Gem5Connection(TelnetConnection):
+
+    def __init__(self,
+                 platform,
+                 host=None,
+                 username=None,
+                 password=None,
+                 port=None,
+                 timeout=None,
+                 password_prompt=None,
+                 original_prompt=None,
+                 ):
+        # TODO: need to insert check that host is the actual place
+        # where devlib is now running
+        self.host = host
+        if username is not None or username is not 'root':
+            gem5_logger.info("User is always root in gem5")
+        if password is not None or password is not '':
+            gem5_logger.info("No password needed in gem5")
+        self.username =  'root'
+        self.password = None
+        self.port = -1
+        # Long timeouts to account for gem5 being slow
+        self.default_timeout = 3600
+        self.ready_timeout = self.default_timeout * 3
+        # Intermediate directory to push things to gem5 using VirtIO
+        self.gem5interactdir = None
+        # Directory to store things from gem5 on the host
+        self.gem5outdir = None
+        self.busybox_wrapper = None
+        self.m5_path = None
+        # Actual telnet connection to gem5 simulation
+        self.conn = None
+        # Flag to indicate the gem5 device is ready to interact with the
+        # outer world
+        self.ready = False
+
+        # Connect to gem5
+        self._connect_gem5(platform)
+
+        # Wait for boot
+        self._wait_for_boot()
+
+        # Mount the virtIO to transfer files in/out gem5 system
+        self._mount_virtio()
+
+        # Ready for interaction now
+        self.ready = True
+
+
+    def set_busybox_wrapper(self, wrapper):
+        self.busybox_wrapper = wrapper
+
+    def connected_as_root(self):
+        # gem5 is always rooted
+        return True
+
+    def push(self, source, dest, timeout=None, wrap_busybox=False):
```

Comment:
Does `wrap_busybox` need to be passed to this function?
This needs to be updated in android adb connection as well if required.